internal/crypto/cms/errors.go

```diff
@@ -0,0 +1,62 @@
+package cms
+
+import "errors"
+
+// ErrExpectSignedData is returned if wrong content is provided when signed
+// data is expected.
+var ErrExpectSignedData = errors.New("cms: signed data expected")
+
+// ErrAttributeNotFound is returned if attribute is not found in a given set.
+var ErrAttributeNotFound = errors.New("attribute not found")
+
+// Verification errors
+var (
+	ErrSignerNotFound      = VerificationError{Message: "signer not found"}
+	ErrCertificateNotFound = VerificationError{Message: "certificate not found"}
+)
+
+// SyntaxError indicates that the ASN.1 data is invalid.
+type SyntaxError struct {
+	Message string
+	Detail  error
+}
+
+// Error returns error message.
+func (e SyntaxError) Error() string {
+	msg := "cms: syntax error"
+	if e.Message != "" {
+		msg += ": " + e.Message
+	}
+	if e.Detail != nil {
+		msg += ": " + e.Detail.Error()
+	}
+	return msg
+}
+
+// Unwrap returns the internal error.
+func (e SyntaxError) Unwrap() error {
+	return e.Detail
+}
+
+// VerificationError indicates verification failures.
+type VerificationError struct {
+	Message string
+	Detail  error
+}
+
+// Error returns error message.
+func (e VerificationError) Error() string {
```

Comment:
The codes look similar. Is it possible to avoid the duplicated code?

Reply:
They are two types of error. Technically, we can avoid the duplicated code but will make readability worse.

Actually, similar pattern can be found in the [go built-in library code](https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/encoding/asn1/asn1.go;l=34-47).